### PR TITLE
Changes required to build whole MB source

### PR DESCRIPTION
**Pull request by DoDoENT**

### CODE REVIEW

File: toolchains/gcc_compatibles.cmake [hunk 1]
```diff
@@ -29,7 +29,7 @@ add_compile_options( -fstrict-aliasing -fstrict-enums -fvisibility=hidden -fvisi
 # "Unknown language" error with CMake 3.5.2 if COMPILE_LANGUAGE:C is used.
 # + 'COMPILE_LANGUAGE' isn't supported by VS generators:
 # https://cmake.org/cmake/help/latest/manual/cmake-generator-expressions.7.html#logical-expressions
-add_compile_options( $<$<COMPILE_LANGUAGE:CXX>:-std=gnu++1z> )
+add_compile_options( $<$<COMPILE_LANGUAGE:CXX>:-std=gnu++14> )
```
Comment (psiha) [inline toolchains/gcc_compatibles.cmake]: why?

File: build_options.cmake [hunk 2]
```diff
@@ -9,10 +9,12 @@
 # http://www.cmake.org/Wiki/CMake_Cross_Compiling
 # https://cmake.org/cmake/help/latest/manual/cmake-toolchains.7.html
 # http://stackoverflow.com/questions/12802377/in-cmake-how-can-i-find-the-directory-of-an-included-file
-if ( WIN32 )
+if ( WIN32 ) # TODO: add detection of Windows Phone / Windows Universal platform
```
Comment (psiha) [inline build_options.cmake:12]: that would probably go with an explicit toolchain file (like android and ios)...

**Comment by psiha** [comment]
Thanks DoDo for the help and cooperation, unfortunately I cannot accept this PR as is since it contains MB specific changes...I will go through those in the commits and mark/comment them...after they are removed (really these should go into your specific build scripts) I'll merge the PR...

File: toolchains/osx.toolchain.cmake [hunk 3]
```diff
@@ -39,17 +39,40 @@ set( TNUN_os_suffix OSX )
 
 set( TNUN_cpu_archs default )
 
+set( TNUN_ABI x64 )
```
Comment (psiha) [inline toolchains/osx.toolchain.cmake:41]: where do you need this?

File: toolchains/osx.toolchain.cmake [hunk 4]
```diff
@@ -39,17 +39,40 @@ set( TNUN_os_suffix OSX )
...
 
-#...mrmlj...todo...
-#if ( <arch> STREQUAL sse3 )
-#    set( XCODE_ATTRIBUTE_CFLAGS_i386   "-msse3  -march=prescott -mtune=core2"  )
-#    set( XCODE_ATTRIBUTE_CFLAGS_x86_64 "-mssse3 -march=core2    -mtune=corei7" )
-#elseif ( <arch> STREQUAL sse4.1 )
-#    set( XCODE_ATTRIBUTE_CFLAGS_i386   "-msse4.1 -march=core2 -mtune=core2"  )
-#    set( XCODE_ATTRIBUTE_CFLAGS_x86_64 "-msse4.1 -march=core2 -mtune=corei7" )
-#endif()
+if("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang")
```
Comment (psiha) [inline toolchains/osx.toolchain.cmake]: this whole block can only go in if it is enabled only for generators other than xcode (as it makes no sense for xcode)
- the MACOSX_CPP_LIBRARY\* variables can be named more generic like TNUN_CPP_LIB
- the clang+stdc++ and gcc+libc++ combinations and their permutations (if we really need them) can be made more generic (you can have the same situation on linux or even windows)...
- i would rather avoid the introduction of USING_CLANG and USING_GCC variables if possible

File: toolchains/msvc.toolchain.cmake [hunk 5]
```diff
@@ -26,20 +28,25 @@ set( TNUN_compiler_exceptions_off        -wd4577
...
 # w4373: '...': virtual function overrides '...', previous versions of the compiler did not override when parameters only differed by const/volatile qualifiers
 # w4324: 'structure was padded due to alignment specifier'
-add_compile_options( /std:c++latest -MP -Oi -W4 -Zc:threadSafeInit- -wd4324 -wd4373 -wd4100 -wd4456 -wd4067 -wd4995 -wd4635 -wd4800 -wd4101 -wd4756 -wd4068 -wd4056 -wd4996 -wd4503)
+add_compile_options( /std:c++latest -MP -Oi -Zc:threadSafeInit- )
 add_definitions(
   -D_CRT_SECURE_NO_WARNINGS
   -D_SCL_SECURE_NO_WARNINGS
   -D_SBCS
   -D_WIN32_WINNT=0x0601 # Win7
+  -DNO_STRICT
```
Comment (psiha) [inline toolchains/msvc.toolchain.cmake]: this is known to break some software/libraries that depend on these...but ok let's leave it like that for now..

File: toolchains/msvc.toolchain.cmake [hunk 6]
```diff
@@ -26,20 +28,25 @@ set( TNUN_compiler_exceptions_off        -wd4577
...
+set( TNUN_disabled_warnings              -wd4324 -wd4373 -wd4100 -wd4456 -wd4067 -wd4995 -wd4635 -wd4800 -wd4101 -wd4756 -wd4068 -wd4056 -wd4996 -wd4503 )
+set( TNUN_warnings_as_errors             -WX )
+set( TNUN_default_warnings               -W4 )
 
-set( TNUN_compiler_release_flags -DNDEBUG -Bt -Ox -Ob2 -Oy -GF -Gw -Gm- -GS- -Gy )
 
 # w4373: '...': virtual function overrides '...', previous versions of the compiler did not override when parameters only differed by const/volatile qualifiers
 # w4324: 'structure was padded due to alignment specifier'
-add_compile_options( /std:c++latest -MP -Oi -W4 -Zc:threadSafeInit- -wd4324 -wd4373 -wd4100 -wd4456 -wd4067 -wd4995 -wd4635 -wd4800 -wd4101 -wd4756 -wd4068 -wd4056 -wd4996 -wd4503)
+add_compile_options( /std:c++latest -MP -Oi -Zc:threadSafeInit- )
```
Comment (psiha) [inline toolchains/msvc.toolchain.cmake:37]: why did you remove the default /W4? i consider this a feature of the build system to default to the 'compiler is your friend' attitude...
what is the reason for the introduction of TNUN_disabled_warnings?

File: toolchains/linux/x86.abi.cmake [hunk 7]
```diff
@@ -0,0 +1,10 @@
+################################################################################
+#
+# T:N.U.N. Linux ABI config file.
+#
+# Copyright (c) 2016. Nenad Miksa. All rights reserved.
+#
+################################################################################
+
+add_compile_options( -m32 )
+link_libraries( -m64 )
```
Comment (psiha) [inline toolchains/linux/x86.abi.cmake]: m64?

File: toolchains/linux.toolchain.cmake [hunk 8]
```diff
@@ -0,0 +1,57 @@
+################################################################################
+#
+# T:N.U.N. Apple OS X CMake tool chain file.
+#
+# Copyright (c) 2016. Domagoj Saric. All rights reserved.
+#
+################################################################################
+
+# MACOSX_RPATH
```
Comment (psiha) [inline toolchains/linux.toolchain.cmake:9]: leftovers? :)

File: toolchains/linux.toolchain.cmake [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+set( TNUN_os_suffix Linux )
+
+if("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang")
+    include( "${CMAKE_CURRENT_LIST_DIR}/clang.cmake" )
+else()
+    include( "${CMAKE_CURRENT_LIST_DIR}/gcc_compatibles.cmake" )
+endif()
+
+option( TNUN_STATIC_LIBC "Link all binaries statically with libc and libc++" false )
```
Comment (psiha) [inline toolchains/linux.toolchain.cmake]: preferably this would be a more generic option (e.g. also often used on windows)...but ok we can add this in a separate change..

File: toolchains/ios.toolchain.cmake [hunk 10]
```diff
@@ -20,6 +20,8 @@
 
 include( "${CMAKE_CURRENT_LIST_DIR}/apple.cmake" )
 
+add_compile_options( $<$<COMPILE_LANGUAGE:CXX>:-x objective-c++> )
```
Comment (psiha) [inline toolchains/ios.toolchain.cmake]: this is a hack...which is sometimes necessary but then one should do it for a specific file in his own build scripts...this cannot go in as a default....that's what the .mm file extension is for...

File: toolchains/gcc_compatibles.cmake [hunk 11]
```diff
@@ -22,9 +23,16 @@ set( TNUN_compiler_exceptions_off      -fno-exceptions                       )
 set( TNUN_compiler_optimize_for_speed  -O3                                   )
 set( TNUN_compiler_optimize_for_size   -Os                                   )
 set( TNUN_compiler_report_optimization -ftree-vectorizer-verbose=6           )
-set( TNUN_compiler_release_flags       -DNDEBUG -fomit-frame-pointer -ffunction-sections -fdata-sections -fmerge-all-constants -fno-stack-protector )
+set( TNUN_compiler_release_flags       -DNDEBUG -fomit-frame-pointer -ffunction-sections -fmerge-all-constants -fno-stack-protector )
+if( ${CMAKE_SYSTEM_NAME} MATCHES "Linux" )
```
Comment (psiha) [inline toolchains/gcc_compatibles.cmake]: why not put this into the linux toolchain?

File: toolchains/gcc_compatibles.cmake [hunk 12]
```diff
@@ -22,9 +23,16 @@ set( TNUN_compiler_exceptions_off      -fno-exceptions                       )
...
+    set( TNUN_linker_release_flags         -Wl,--gc-sections                     )
+endif()
+set( TNUN_default_warnings             -Wall -Wextra -Wstrict-aliasing       )
+set( TNUN_warnings_as_errors           -Werror                               )
+set( TNUN_disabled_warnings            -Wno-unknown-pragmas -Wno-unused-parameter -Wno-ignored-qualifiers -Wno-sign-conversion -Wno-clobbered -Wno-format-zero-length -Wno-error=deprecated-declarations     )
+set( TNUN_native_optimization          -march=native -mtune=native           )
 
-add_compile_options( -fstrict-aliasing -fstrict-enums -fvisibility=hidden -fvisibility-inlines-hidden -fno-threadsafe-statics -Wall -Wstrict-aliasing -Wno-multichar -Wno-unknown-pragmas -Wno-unused-local-typedefs )
+add_compile_options( -fstrict-aliasing -fstrict-enums -fvisibility=hidden -fvisibility-inlines-hidden -fPIC )
```
Comment (psiha) [inline toolchains/gcc_compatibles.cmake]: where did  -fno-threadsafe-statics go?

File: toolchains/gcc_compatibles.cmake [hunk 13]
```diff
@@ -22,9 +23,16 @@ set( TNUN_compiler_exceptions_off      -fno-exceptions                       )
...
 set( TNUN_compiler_optimize_for_size   -Os                                   )
 set( TNUN_compiler_report_optimization -ftree-vectorizer-verbose=6           )
-set( TNUN_compiler_release_flags       -DNDEBUG -fomit-frame-pointer -ffunction-sections -fdata-sections -fmerge-all-constants -fno-stack-protector )
+set( TNUN_compiler_release_flags       -DNDEBUG -fomit-frame-pointer -ffunction-sections -fmerge-all-constants -fno-stack-protector )
+if( ${CMAKE_SYSTEM_NAME} MATCHES "Linux" )
+    set( TNUN_linker_release_flags         -Wl,--gc-sections                     )
+endif()
+set( TNUN_default_warnings             -Wall -Wextra -Wstrict-aliasing       )
+set( TNUN_warnings_as_errors           -Werror                               )
+set( TNUN_disabled_warnings            -Wno-unknown-pragmas -Wno-unused-parameter -Wno-ignored-qualifiers -Wno-sign-conversion -Wno-clobbered -Wno-format-zero-length -Wno-error=deprecated-declarations     )
```
Comment (psiha) [inline toolchains/gcc_compatibles.cmake]: where do you have problems with Wclobbered?
Wunused-parameter is not a trivial warning and i don't want it disabled by default...but again why the introduction of this TNUN_disabled_warnings variable?

File: toolchains/gcc_compatibles.cmake [hunk 14]
```diff
@@ -9,9 +9,10 @@
 # https://cmake.org/Bug/view.php?id=15939
 # http://stackoverflow.com/questions/31355692/cmake-support-for-gccs-link-time-optimization-lto
 set( CMAKE_INTERPROCEDURAL_OPTIMIZATION true )
-set( CMAKE_VISIBILITY_INLINES_HIDDEN    true )
+#set( CMAKE_VISIBILITY_INLINES_HIDDEN    true )
```
Comment (psiha) [inline toolchains/gcc_compatibles.cmake]: ?

File: toolchains/gcc.cmake [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+#
+# Copyright (c) 2016. Nenad Miksa. All rights reserved.
+#
+################################################################################
+
+include( "${CMAKE_CURRENT_LIST_DIR}/gcc_compatibles.cmake" )
+
+list( APPEND TNUN_disabled_warnings -Wno-error=cpp -Wno-deprecated-declarations -Wno-unused-result )
+
+if( ${CMAKE_SYSTEM_NAME} MATCHES "Linux" )
```
Comment (psiha) [inline toolchains/gcc.cmake:11]: why just linux?

File: toolchains/clang.cmake [hunk 16]
```diff
@@ -12,9 +12,16 @@ include( "${CMAKE_CURRENT_LIST_DIR}/gcc_compatibles.cmake" )
 
 list( APPEND TNUN_compiler_optimize_for_speed -fvectorize -fslp-vectorize -fslp-vectorize-aggressive )
 list( APPEND TNUN_compiler_report_optimization -Rpass=loop-.* )
+list( APPEND TNUN_disabled_warnings -Wno-unknown-warning-option "-Wno-error=#warnings" -Wno-attributes $<$<COMPILE_LANGUAGE:CXX>:-Wno-overloaded-virtual> )
 
 # http://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html
 # http://clang.llvm.org/docs/UsersManual.html#controlling-code-generation
-set( TNUN_compiler_runtime_sanity_checks -fsanitize=undefined -fsanitize=integer )
+set( TNUN_compiler_runtime_sanity_checks -fsanitize=undefined -fsanitize=integer -fsanitize=address -fno-omit-frame-pointer )
```
Comment (psiha) [inline toolchains/clang.cmake:19]: why do you need -fno-omit-frame-pointer here?

File: toolchains/clang.cmake [hunk 17]
```diff
@@ -12,9 +12,16 @@ include( "${CMAKE_CURRENT_LIST_DIR}/gcc_compatibles.cmake" )
...
 
-add_compile_options( -Wheader-guard )
+if( "${CMAKE_CXX_COMPILER}" MATCHES ".*ccache" )
```
Comment (psiha) [inline toolchains/clang.cmake:34]: can you add an implementation note explaining this block?

File: build_options.cmake [hunk 18]
```diff
@@ -52,8 +54,27 @@ function( TNUN_add_compile_options configuration )
...
+else()
+    add_compile_options( ${TNUN_compiler_exceptions_off} )
+endif()
+
+option( TNUN_ALLOW_RTTI "Allow Runtime Type Information in C++ code" true )
+if( ${TNUN_ALLOW_RTTI} )
+    add_compile_options( ${TNUN_compiler_rtti_on} )
+else()
+    add_compile_options( ${TNUN_compiler_rtti_off} )
+endif()
```
Comment (psiha) [inline build_options.cmake]: not to happy about this block: in ide's/multiconfig generators you'll often want rtti in debug but not in release and also, although that is less common, the same for exceptions..

File: toolchains/gcc_compatibles.cmake (hunk 1 above)
Comment (DoDoENT) [inline toolchains/gcc_compatibles.cmake]: Qt 4.7 and PCH compatibility. When linking with targets created by Qt framework, they automatically publish compiler flag `-std=gnu++11` or `-std=gnu++14` to your code. OTOH, when assembling flags for PCH, this flag may still not be published (e.g. in cases target for which PCH is added depends on target which is not yet defined but which ultimately depends on Qt - PCH then ends up with `-std=gnu++1z`, whilst code ends up with `-std=gnu++14`. AFAIK, this issue can only be solved by CMake adding truly native support for PCH (without hacks we currently use).

File: build_options.cmake (hunk 2 above)
Comment (DoDoENT) [inline build_options.cmake:12]: Not really - WP uses default MSVC toolchain.

For example, to configure code for WP 8.0 ARM, you should use:

```
cmake -G "Visual Studio 11 2012 ARM" -DCMAKE_SYSTEM_NAME=WindowsPhone -DCMAKE_SYSTEM_VERSION=8.0
```

For WP 8.1, similarly:

```
cmake -G "Visual Studio 12 2013 ARM" -DCMAKE_SYSTEM_NAME=WindowsPhone -DCMAKE_SYSTEM_VERSION=8.1
```

And for Windows Mobile 10.0 / Universal Windows App:

```
cmake -G "Visual Studio 14 2015 ARM" -DCMAKE_SYSTEM_NAME=WindowsStore -DCMAKE_SYSTEM_VERSION=10.0
```

Of course, specifiyng `Win64` instead of `ARM` will generate 64-bit VS project and not specifying anything will generate 32-bit (x86) VS project. On WindowsPhone, `x86` is usually interpreted as simulator build, while WindowsStore can be either `x86`, `ARM` or `Win64` (in Windows 8.0 and 8.1, WindowStore referred to Metro apps which used different MSVCRT than WindowsPhone - on Windows 10, WindowsStore system name refers to universal windows app).

File: toolchains/osx.toolchain.cmake (hunk 3 above)
Comment (DoDoENT) [inline toolchains/osx.toolchain.cmake:41]: In MB Source when we give names to tests, we add ABI suffix to be able to distinguish them in CI server report. In old build system this was PHOTOPAY_ARCHITECTURE variable which is now TNUN_ABI. On MacOS we always have x64 ABI.

Generally, this can be removed from build, but that would then require us to have parallel osx.toolchain.cmake file that would have to be loaded after this one.

File: toolchains/osx.toolchain.cmake (hunk 4 above)
Comment (DoDoENT) [inline toolchains/osx.toolchain.cmake]: USING_CLANG and USING_GCC are for simplicity (to make following ifs easier to read) and can be removed.

I agree about TNUN_CPP_LIB variable which should generally be platform agnostic, but OTOH forcing to use specific STL with specific compiler requires flags specific to both compiler and platform (see line 71). This is something that should probably be defined in separate "toolchain add-on" files.

File: toolchains/msvc.toolchain.cmake (hunk 6 above)
Comment (DoDoENT) [inline toolchains/msvc.toolchain.cmake:37]: We treat all warnings as errors and W4 introduces some very silly warnings which would then either need to be disabled with TNUN_disabled_warnings or fixed (in our source this would mean several weeks of work).

TNUN_disabled_warnings variable is added to contain silly warnings that should be disabled because they are both unimportant and difficult or impossible to fix. We treat other warnings as errors so adding those would make code uncompilable. AFAIK, MSVC does not have the ability to treat all warnings as errors, except some specifically defined (something like -Wno-error= on GCC/Clang). Even in that case, producing a lot of warnings, especially in third party libs, would produce a lot of noise in compile output, making it difficult for developers to find actual compile error they are facing (not everyone is using VS which separates warnings from errors in GUI).

File: toolchains/linux.toolchain.cmake (hunk 8 above)
Comment (DoDoENT) [inline toolchains/linux.toolchain.cmake:9]: yep

File: toolchains/gcc_compatibles.cmake (hunk 11 above)
Comment (DoDoENT) [inline toolchains/gcc_compatibles.cmake]: The original idea was to also have this flag on android, but since it was added in android toolchain, it can be removed from here.

File: toolchains/gcc_compatibles.cmake (hunk 13 above)
Comment (DoDoENT) [inline toolchains/gcc_compatibles.cmake]: -Wunused-parameter is very annoying when overriding virtual functions and not using all parameters.

-Wno-clobbered is probably leftover from some legacy code or support for compiling some third party libraries we use in PhotoPay (kontocheck would be my first guess) - I'll have to recheck whether everyhing works without it.

File: toolchains/gcc_compatibles.cmake (hunk 14 above)
Comment (DoDoENT) [inline toolchains/gcc_compatibles.cmake]: Protobuf has problems when this is enabled - macOS linker throws lots of warnings.

File: toolchains/clang.cmake (hunk 17 above)
Comment (DoDoENT) [inline toolchains/clang.cmake:34]: When clang is used behind ccache, it throws a lot of "unused-argument" warnings. I'm not sure why that happens (it also happens on clang for Android, both from ndk-build and cmake android build). Having this much warnings for each and every source file is very annoying.

File: toolchains/gcc_compatibles.cmake (hunk 1 above)
Comment (psiha) [inline toolchains/gcc_compatibles.cmake]: If we both agree to push for "/std:latest", can you find another way to hack around this? E.g. remove this option afterwards with remove_definitions or override it with another add_compile_options (or a directory or target specific version of it)?

File: toolchains/osx.toolchain.cmake (hunk 3 above)
Comment (psiha) [inline toolchains/osx.toolchain.cmake:41]: hm yes...universal build don't play well with this scheme...
so let's leave it like that for now...and later we can then add separate x86 and x86+x64 'unibuild' abis if needed...

File: toolchains/osx.toolchain.cmake (hunk 4 above)
Comment (psiha) [inline toolchains/osx.toolchain.cmake]: since there is the builtin MSVC i'd rather go with CLANG and GCC then..

File: toolchains/msvc.toolchain.cmake (hunk 6 above)
Comment (psiha) [inline toolchains/msvc.toolchain.cmake:37]: adding /w3 afterwards should acomplish what you wish?
and /wd should work the same as -Wno-error?

File: toolchains/gcc_compatibles.cmake (hunk 13 above)
Comment (psiha) [inline toolchains/gcc_compatibles.cmake]: no go @ -Wunused-parameter, that's really the basics in tidy code ... simply don't give the parameter a name (or put its name in a /**/ comment block...standard practice)...+ plus you can add that option locally..

File: toolchains/gcc_compatibles.cmake (hunk 14 above)
Comment (psiha) [inline toolchains/gcc_compatibles.cmake]: hidden visibility should really be standard practice these days...can't you disable it just for protobuf?

File: toolchains/gcc.cmake (hunk 15 above)
Comment (psiha) [inline toolchains/gcc.cmake:11]: ?

File: toolchains/clang.cmake (hunk 16 above)
Comment (psiha) [inline toolchains/clang.cmake:19]: ?

File: toolchains/clang.cmake (hunk 17 above)
Comment (psiha) [inline toolchains/clang.cmake:34]: ok but can you add that as an implementation note inside the clang.cmake file?

File: build_options.cmake [hunk 19]
```diff
@@ -14,7 +14,9 @@ if ( WIN32 ) # TODO: add detection of Windows Phone / Windows Universal platform
 elseif( APPLE AND NOT iOS )
     include( "${CMAKE_CURRENT_LIST_DIR}/toolchains/osx.toolchain.cmake" )
 elseif( ${CMAKE_SYSTEM_NAME} MATCHES "Linux" )
-    include ( "${CMAKE_CURRENT_LIST_DIR}/toolchains/linux.toolchain.cmake" )
+    include( "${CMAKE_CURRENT_LIST_DIR}/toolchains/linux.toolchain.cmake" )
+elseif( ANDROID_TOOLCHAIN ) # TNUN android.toolchain.cmake does not define this variable, while native Android Studio toolchain does
```
Comment (psiha) [inline build_options.cmake:18]: aren't you essentially including the android toolchin file twice here?

File: build_options.cmake [hunk 20]
```diff
@@ -63,17 +65,39 @@ if ( ${TNUN_DEBUG_SYMBOLS_IN_RELEASE} )
 endif()
 
 option( TNUN_ALLOW_EXCEPTIONS "Allow exception support in C++ code" true )
-if( ${TNUN_ALLOW_EXCEPTIONS} )
-    add_compile_options( ${TNUN_compiler_exceptions_on} )
+option( TNUN_ALLOW_RTTI "Allow Runtime Type Information in C++ code" false )
+
+if( ${CMAKE_GENERATOR} MATCHES "Visual Studio" )
```
Comment (psiha) [inline build_options.cmake]: huh...why did this explode?

File: toolchains/android-studio.toolchain.cmake [hunk 21]
```diff
@@ -0,0 +1,66 @@
+################################################################################
+#
+# T:N.U.N. Android Studio CMake tool chain file. This is to be used together with
+# native android CMake toolchain bundled with Android studio 2.2 and newer.
+# For uses without Android Studio, please use android.toolchain.cmake file.
```
Comment (psiha) [inline toolchains/android-studio.toolchain.cmake:5]: can't these two toolchains be merged into one?

File: toolchains/android.toolchain.cmake [hunk 22]
```diff
@@ -228,6 +228,7 @@ set( CMAKE_OBJDUMP      "${ANDROID_TOOLCHAIN_ROOT}/bin/${_CMAKE_TOOLCHAIN_PREFIX
 set( CMAKE_RANLIB       "${ANDROID_TOOLCHAIN_ROOT}/bin/${_CMAKE_TOOLCHAIN_PREFIX}ranlib${TOOL_OS_SUFFIX}"  )
 
 include( "${CMAKE_CURRENT_LIST_DIR}/clang.cmake" )
+unset( TNUN_native_optimization ) # This makes no sense when cross-compiling.
```
Comment (psiha) [inline toolchains/android.toolchain.cmake:231]: why not simply not-give the option when crosscompiling in the first place?

File: toolchains/apple.cmake [hunk 23]
```diff
@@ -41,9 +41,9 @@ set( TNUN_ABI   default    )
 set( TNUN_ABIs ${TNUN_ABI} )
 
 #...mrmlj...reinvestigate this...
-set( CMAKE_XCODE_ATTRIBUTE_OBJROOT          "${PROJECT_BINARY_DIR}" )
-set( CMAKE_XCODE_ATTRIBUTE_BUILD_DIR        "${PROJECT_BINARY_DIR}" )
-set( CMAKE_XCODE_ATTRIBUTE_BUILD_ROOT       "${PROJECT_BINARY_DIR}" )
-set( CMAKE_XCODE_ATTRIBUTE_PROJECT_TEMP_DIR "${PROJECT_BINARY_DIR}" )
-set( CMAKE_XCODE_ATTRIBUTE_SYMROOT          "${PROJECT_BINARY_DIR}/lib" )
-set( CMAKE_XCODE_ATTRIBUTE_SYMROOT_RELEASE  "${PROJECT_BINARY_DIR}/lib" )
+# set( CMAKE_XCODE_ATTRIBUTE_OBJROOT          "${PROJECT_BINARY_DIR}" )
```
Comment (psiha) [inline toolchains/apple.cmake:51]: you have give a reason for commenting things out?

File: toolchains/msvc.toolchain.cmake [hunk 24]
```diff
@@ -30,7 +30,7 @@ set( TNUN_compiler_optimize_for_speed    -Ox -Ot -Qpar
 set( TNUN_compiler_runtime_sanity_checks -GS -sdl -guard:cf -fp:strict -RTC1 -RTCc -D_ALLOW_RTCc_IN_STL ) # https://www.reddit.com/r/cpp/comments/46mhne/rtcc_rejects_conformant_code_with_visual_c_2015
 set( TNUN_disabled_warnings              -wd4324 -wd4373 -wd4100 -wd4456 -wd4067 -wd4995 -wd4635 -wd4800 -wd4101 -wd4756 -wd4068 -wd4056 -wd4996 -wd4503 )
 set( TNUN_warnings_as_errors             -WX )
-set( TNUN_default_warnings               -W4 )
+set( TNUN_default_warnings               -W3 )
```
Comment (psiha) [inline toolchains/msvc.toolchain.cmake]: /W4 has to be the default...

File: toolchains/msvc.toolchain.cmake [hunk 25]
```diff
@@ -64,6 +64,21 @@ endif()
 set( TNUN_arch_include_dir "${CMAKE_CURRENT_LIST_DIR}/windows" )
 include( "${TNUN_arch_include_dir}/${TNUN_ABI}.abi.cmake" )
 
+# Detect Visual Studio Version
+if( ${CMAKE_GENERATOR} MATCHES "Visual Studio" )
+    if(${CMAKE_CXX_COMPILER_VERSION} VERSION_LESS 20 AND ${CMAKE_CXX_COMPILER_VERSION} VERSION_GREATER 19)
```
Comment (psiha) [inline toolchains/msvc.toolchain.cmake]: why do you need this?

File: toolchains/osx.toolchain.cmake [hunk 26]
```diff
@@ -17,10 +17,11 @@ include( "${CMAKE_CURRENT_LIST_DIR}/apple.cmake" )
 cmake_policy( SET CMP0042 OLD )
 set( CMAKE_MACOSX_RPATH 0 ) # even with the policy set CMake 3.5.2 still issues the warning?
 
-set( CMAKE_OSX_ARCHITECTURES           "$(ARCHS_STANDARD_32_64_BIT)" )
+set( CMAKE_OSX_ARCHITECTURES           "$(ARCHS_STANDARD)" )
 set( CMAKE_XCODE_ATTRIBUTE_VALID_ARCHS "$(ARCHS_STANDARD_32_64_BIT)" )
-set( CMAKE_OSX_SYSROOT                 "macosx"                      ) #"Latest Mac OS X"
-set( CMAKE_OSX_DEPLOYMENT_TARGET       "10.7"                        )
+if( ${CMAKE_GENERATOR} MATCHES "Xcode" )
+    set( CMAKE_OSX_DEPLOYMENT_TARGET "10.11"                       )
```
Comment (psiha) [inline toolchains/osx.toolchain.cmake]: no can do...other people/projects still have to support older osx versions...
10.7 has libcpp by default and that is enough as a base minimum...
and why the condition on xcode?

File: toolchains/gcc_compatibles.cmake (hunk 1 above)
Comment (DoDoENT) [inline toolchains/gcc_compatibles.cmake]: I'll try to. However, the real fix for this issue would be for CMake guys to fix [issue 1260](https://gitlab.kitware.com/cmake/cmake/issues/1260).

File: toolchains/osx.toolchain.cmake (hunk 26 above)
Comment (DoDoENT) [inline toolchains/osx.toolchain.cmake]: I agree on 10.7 - I've put 10.11 because MB 3rd party libs are built with 10.11 min target and causes linker warnings. I agree this should be overriden in our source.

Setting `CMAKE_OSX_DEPLOYMENT_TARGET` on makefile/ninja generator causes problems with PCH.

File: toolchains/clang.cmake (hunk 16 above)
Comment (DoDoENT) [inline toolchains/clang.cmake:19]: To get nicer stack trace in sanitization report. See [documentation](http://clang.llvm.org/docs/AddressSanitizer.html#usage)

File: build_options.cmake (hunk 19 above)
Comment (DoDoENT) [inline build_options.cmake:18]: Essentially yes. Android Studio first forces its own android toolchain which does not have as optimal flags as TNUN android cmake. Since you cannot force AS to use non-default toolchain, this includes a special "android-studio toolchain" which improves flags imposed by AS default toolchain.

**Comment by psiha** [comment]
Google is your friend ;P
https://blogs.msdn.microsoft.com/vcblog/2016/06/07/standards-version-switches-in-the-compiler
#define _HAS_AUTO_PTR_ETC 1
(I actually though this was already in the build somewhere...)

I'd like to document this build system as 'by default' promoting latest C++ practices and language advancements, Boost, high warning levels etc...
(and yes for now it requires MSVC15u3/latest VS, latest Xcode)...
